### PR TITLE
[MLX004I] E01: Automated Quality Framework

### PR DESCRIPTION
**Pull request by ronniepinnell**
## Summary
Build a reusable, automated quality gate that runs per-source checks on canonical training data and produces PASS/FAIL reports. Integrated into training pipeline as a hard gate.

## Epic
BEN-1980: [MLX004I] E01: Automated Quality Framework

## Tasks Included
- BEN-1990: Build quality_gate.py — per-source quality check framework (6 checks: schema, nulls, ordering, duplicates, distributions, features)
- BEN-1991: Log quality gate results to Supabase qa.training_quality_runs (migration + --log-to-supabase flag)
- BEN-1992: Integrate quality gate into training pipeline — refuse to train if gate fails (--skip-quality-gate escape hatch)

## Test Results
```
tests/milestones/mlx004i/test_e01_quality_framework.py — 6/6 PASSED (78s)
  ✓ test_script_exists
  ✓ test_importable
  ✓ test_runs_per_source_checks
  ✓ test_produces_json_report
  ✓ test_nonzero_exit_on_failure
  ✓ test_cli_all_flag
```

## Quality Gate Results (current state)
- **PASS:** bdc, halo, hockey_stats_nhl, kaggle, moneypuck, nhl_api, pwhl (7/8)
- **FAIL:** a3z (6.6% null UUIDs, ordering violations) — known issue for E03

## Acceptance Criteria
- [x] quality_gate.py exists and is importable
- [x] Runs per-source checks: ordering, nulls, dupes, schema, distributions, features
- [x] Produces JSON report per source
- [x] Returns non-zero exit code if any check fails
- [x] Results written to Supabase via --log-to-supabase flag (migration included)
- [x] train_multi_source.py calls quality gate before training

---

## Summary by CodeRabbit

* **New Features**
  * Introduced a standalone data quality CLI that runs multiple automated checks (schema, nulls, ordering, duplicates, distributions, numeric sanity) on canonical training data, with console summaries and optional JSON export.
  * Added an option to log quality runs to a remote store and persisted run results for auditing.
  * Added a pre-training quality gate (with --skip-quality-gate) that halts training on failures.

* **Tests**
  * Added acceptance tests validating the quality tool, its outputs, and CLI behavior.

### CODE REVIEW

**Comment by ronniepinnell** [comment]
## Code Review — PR #9358 (Quality Gate Framework)

### CRITICAL BUGS

**1. `NameError` in `train_multi_source.py` lines 1048 and 1058 — variables used before assignment**

After `build_sequences()` returns `X_all, y_all`, the code immediately references `X_train`, `X_val`, `X_test` which don't exist yet (they're created at lines 1063-1068):

```python
# Line 1048 — X_train/X_val/X_test don't exist yet
if len(X_train) + len(X_val) + len(X_test) == 0:

# Line 1058 — same problem
indices = list(range(len(X_train) + len(X_val) + len(X_test)))
```

Both should use `len(X_all)` instead. This will crash on every cache-miss run.

**2. Schema mismatch between Python and SQL — `quality_gate.py` line 354**

The migration creates `qa.training_quality_runs`, but the Python code inserts into `training_quality_runs` without specifying the `qa` schema:

```python
client.table("training_quality_runs").insert(rows).execute()
```

This will either fail (table not found in `public` schema) or insert into the wrong table if a `public.training_quality_runs` happens to exist. Fix: use `client.schema("qa").table("training_quality_runs")`.

### BUGS

**3. `df` unbound on cache hit — `train_multi_source.py` lines 1126 and 1197-1199**

When the sequence cache is hit, `df` is never assigned. Line 1126 (`df["source"].unique()`) will throw inside a broad `except Exception` that silently swallows it — the R2 upload metadata will have an empty `data_sources` list. Lines 1197-1199 catch `UnboundLocalError` explicitly, but this is a design smell. Consider loading source info from cache metadata or the canonical directory listing.

**4. Test ordering bug — `test_e01_quality_framework.py` line 59-66**

The skip-if-no-data check runs AFTER the subprocess:

```python
result = subprocess.run(["python3", ..., "--source", "a3z"], ...)
if not (CANONICAL_EVENTS_DIR / "a3z" / "events.parquet").exists():
    pytest.skip("No a3z data available")
assert result.returncode != 0
```

If a3z data doesn't exist, the subprocess exits with returncode 1 (file_exists check fails), so the assertion passes — but it's testing file-not-found, not data quality. The skip should be BEFORE the subprocess call.

### SUGGESTIONS

**5. No RLS policy on `qa.training_quality_runs`** — The migration doesn't enable RLS. If this table should be accessible only to service role (likely), add `ALTER TABLE qa.training_quality_runs ENABLE ROW LEVEL SECURITY;` with appropriate policies. Not blocking if this is internal-only, but worth noting per project conventions.

**6. `check_ordering` uses nested `for` loops over grouped DataFrames** — This violates the project's "never nest Python for-loops over DataFrames" rule. Consider vectorizing with `groupby(["game_uuid", "period"]).apply()` on the diff check.

### VERDICT

Two critical bugs (items 1 and 2) must be fixed before merge. The `quality_gate.py` module itself is well-structured — clean separation of checks, good severity levels, proper CLI interface. The SQL migration is solid with a useful summary view.

**Comment by ronniepinnell** [comment]
Fixed in bb826852b:

1. **Schema qualifier** — added `.schema("qa")` to Supabase insert call
2. **Test ordering** — moved `pytest.skip()` before `subprocess.run()` 
3. **Vectorized check_ordering** — replaced nested for-loops with `groupby().diff()` + boolean masking

Note: Bug #1 (NameError on cache-miss path in train_multi_source.py lines 1029/1039) is **pre-existing** — not introduced by this PR. Will file separately.
Bug #3 (df unbound on cache hit) is also pre-existing.
